assets/src/edit-story/components/richText/provider.js

```diff
@@ -0,0 +1,196 @@
+  const [editorState, setEditorState] = useState(null);
+
+  const [
+    selectionIsBold,
+    selectionIsItalic,
+    selectionIsUnderline,
+  ] = useMemo(() => {
+    if (editorState) {
+      return [
+        isBold(editorState),
+        isItalic(editorState),
+        isUnderline(editorState),
+      ];
+    }
+    return [false, false, false];
+  }, [editorState]);
+
+  const setStateFromContent = useCallback(
+    (content) => {
+      const { offset, clearContent, selectAll } = editingElementState || {};
+      let state = EditorState.createWithContent(customImport(content));
+      if (clearContent) {
+        // If `clearContent` is specified, push the update to clear content so that
+        // it can be undone.
+        state = EditorState.push(state, customImport(''), 'remove-range');
+      }
+      let selection;
+      if (selectAll) {
+        selection = getSelectionForAll(state.getCurrentContent());
+      } else if (offset) {
+        selection = getSelectionForOffset(state.getCurrentContent(), offset);
+      }
+      if (selection) {
+        state = EditorState.forceSelection(state, selection);
+      }
+      setEditorState(state);
+    },
+    [editingElementState, setEditorState]
+  );
+
+  // This is to allow the getContentFromState to be stable and *not*
+  // depend on editorState, as would otherwise be a lint error.
+  const lastKnownState = useRef(null);
+  const lastKnownSelection = useRef(null);
+  useEffect(() => {
+    lastKnownState.current = editorState;
+    if (editorState?.getSelection()?.hasFocus) {
+      lastKnownSelection.current = editorState.getSelection();
+    }
+  }, [editorState]);
+
+  const getContentFromState = useCallback((someEditorState) => {
+    if (!someEditorState) {
+      return null;
+    }
+
+    return customExport(someEditorState)
+      .replace(/<br ?\/?>/g, '')
+      .replace(/&nbsp;$/, '');
+  }, []);
+
+  // This filters out illegal content (see `getFilteredState`)
+  // on paste and updates state accordingly.
+  // Furthermore it also sets initial selection if relevant.
+  const updateEditorState = useCallback(
+    (newEditorState) => {
+      const filteredState = getFilteredState(newEditorState, editorState);
+      setEditorState(filteredState);
+    },
+    [editorState, setEditorState]
+  );
+
+  const getHandleKeyCommand = useCallback(
+    () => getHandleKeyCommandFromState(updateEditorState),
+    [updateEditorState]
+  );
+
+  const clearState = useCallback(() => {
+    setEditorState(null);
+    lastKnownSelection.current = null;
+  }, [setEditorState]);
+
+  const hasCurrentEditor = Boolean(editorState);
+
+  const updateWhileUnfocused = useCallback((updater) => {
+    const oldState = lastKnownState.current;
+    const selection = lastKnownSelection.current;
+    const workingState = EditorState.forceSelection(oldState, selection);
+    const newState = updater(workingState);
+    setEditorState(newState);
+    setForceFocus(true);
```

Comment:
This is part of the magic happening. When you press a button (or use a dropdown, input field or color picker) outside the editor, the editor loses focus, but can remember the selection. Here we simply reapply that selection, then perform the formatting as requested, and then ask the editor to refocus the input field - moving focus away from whatever the user pressed elsewhere.

This however is not going to work with color picker, as the text color should probably update as the user clicks around in the color picker, but **not** move focus away from the color picker, as that will close it.

Similarly for the letter spacing input, as you can type and see it take effect in real time, without losing focus in the letter spacing input field.

Reply:
And this came up now with letter spacing - as soon as you type a single letter, focus moves back to the editor. Will have to exclude force focus for this one.

Reply:
Fixed in 4a28da7d.